src/number_theory/modular.lean

```diff
@@ -371,6 +400,266 @@ begin
+  { rw [zpow_sub, zpow_one, coe_mul, h, coe_T_inv], ext i j,
+    fin_cases i; fin_cases j; simp [matrix.mul, dot_product, fin.sum_univ_succ]; ring, },
+end
+
+/- If `c = 1`, then `g = [[1,a],[0,1]] * S * [[1,d],[0,1]]`. -/
+lemma g_eq_of_c_eq_one (g : SL(2,ℤ)) (hc : ↑ₘg 1 0 = 1) :
+  g = T^(↑ₘg 0 0) * S * T^(↑ₘg 1 1) :=
+begin
+  ext i j, fin_cases i; fin_cases j,
+  { simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ] },
+  { have g_det : (1:ℤ) = ↑ₘg 0 0 * ↑ₘg 1 1 - 1 * ↑ₘg 0 1,
+    { convert det_fin_two ↑ₘg using 1,
+      { rw g.det_coe },
+      rw hc,
+      ring },
+    simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ],
+    rw g_det,
+    simp, },
+  { simpa [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ] using hc },
+  { simp [S, coe_T_zpow, matrix.mul_apply, fin.sum_univ_succ], },
+end
+
+lemma cast_one_le_of_pos {α : Type*} [linear_ordered_ring α] {n : ℤ} (hn : 0 < n) :
+  (1 : α) ≤ n := by exact_mod_cast int.add_one_le_of_lt hn
+
+lemma cast_le_neg_one_of_neg {α : Type*} [linear_ordered_ring α] {n : ℤ} (hn : n < 0) :
+  (n : α) ≤ -1 := by exact_mod_cast int.le_sub_one_of_lt hn
+
+lemma nneg_mul_add_sq_of_abs_le_one {α : Type*} [linear_ordered_ring α] (n : ℤ) (x : α)
+  (hx : |x| ≤ 1) : (0 : α) ≤ n * x + n * n :=
+begin
+  have hnx : 0 < n → 0 ≤ x + n := λ hn, by
+  { convert add_le_add (neg_le_of_abs_le hx) (cast_one_le_of_pos hn),
+    rw add_left_neg, },
+  have hnx' : n < 0 → x + n ≤ 0 := λ hn, by
+  { convert add_le_add (le_of_abs_le hx) (cast_le_neg_one_of_neg hn),
+    rw add_right_neg, },
+  rw [← mul_add, mul_nonneg_iff],
+  rcases lt_trichotomy n 0 with h | rfl | h,
+  { exact or.inr ⟨by exact_mod_cast h.le, hnx' h⟩, },
+  { simp [le_total 0 x], },
+  { exact or.inl ⟨by exact_mod_cast h.le, hnx h⟩, },
+end
+
+/-- Nontrivial lemma: if `|x|<1/2` and `n : ℤ`, then `2nx+n^2 ≥ 0`. (False for `n : ℝ`!) -/
+lemma _root_.int.non_neg_of_lt_half (n : ℤ) (x : ℝ) (hx : |x| < 1/2) : (0:ℝ) ≤ 2 * n * x + n * n :=
+begin
+  convert nneg_mul_add_sq_of_abs_le_one n (2*x) _ using 1,
+  { ring, },
+  rw _root_.abs_mul,
+  norm_num,
+  nlinarith,
+end
+
+/-- If `z∈𝒟ᵒ`, and `n:ℤ`, then `|z+n|>1`. -/
+lemma move_by_T {z : ℍ} (hz : z ∈ 𝒟ᵒ) (n : ℤ) : 1 < norm_sq (((T^n) • z) : ℍ) :=
+begin
+  simp only [coe_T_zpow, upper_half_plane.num, coe_smul, coe_fn_eq_coe, coe_matrix_coe,
+    int.coe_cast_ring_hom, map_apply, cons_val_zero, int.cast_one, of_real_one, one_mul,
+    cons_val_one, head_cons, of_real_int_cast, denom_apply, int.cast_zero, zero_mul, zero_add,
+    div_one, complex.norm_sq_apply],
+  have hz1 : 1 < z.re * z.re + z.im * z.im,
+  { have := hz.1,
+    rw norm_sq at this,
+    convert this using 1, },
+  rw (by simp : ((z:ℂ) + n).im = z.im),
+  rw (by simp : ((z:ℂ) + n).re = z.re + n),
+  rw (by ring : (z.re + ↑n) * (z.re + ↑n) = z.re * z.re + 2 * n * z.re + n * n),
+  have : 0 ≤  2 * ↑n * z.re + ↑n * ↑n := int.non_neg_of_lt_half n (z.re) hz.2,
+  convert add_lt_add_of_le_of_lt this hz1 using 1,
```

Comment:
Now that we have `nneg_mul_add_sq_of_abs_le_one` I think I'd probably drop `_root_.int.non_neg_of_lt_half` entirely and just inline its proof here:
```suggestion
  have : 0 ≤  2 * ↑n * z.re + ↑n * ↑n,
  { convert nneg_mul_add_sq_of_abs_le_one n (2*z.re) _ using 1,
    { ring, },
    rw _root_.abs_mul,
    norm_num,
    have := hz.2,
    nlinarith, },
```